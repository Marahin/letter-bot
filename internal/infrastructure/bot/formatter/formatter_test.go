package formatter

import (
	"errors"
	"testing"
	"time"

	"github.com/gkampitakis/go-snaps/snaps"

	"spot-assistant/internal/core/dto/book"
	"spot-assistant/internal/core/dto/guild"
	"spot-assistant/internal/core/dto/member"
	"spot-assistant/internal/core/dto/reservation"
)

func TestDiscordFormatter_FormatGenericError(t *testing.T) {
	// given
	err := errors.New("test error")
	formatter := NewFormatter()

	// when
	output := formatter.FormatGenericError(err)

	// assert
	snaps.MatchSnapshot(t, output)
}

func TestDiscordFormatter_FormatUnbookResponse(t *testing.T) {
	// given
	formatter := NewFormatter()
	res := &reservation.ReservationWithSpot{
		Reservation: reservation.Reservation{
			ID:              0,
			Author:          "sample-author",
			CreatedAt:       time.Time{},
			StartAt:         time.Date(2021, 1, 1, 0, 0, 0, 0, time.UTC),
			EndAt:           time.Date(2021, 1, 1, 2, 0, 0, 0, time.UTC),
			SpotID:          0,
			GuildID:         "sample-guild-id",
			AuthorDiscordID: "sample-author-discord-id",
		},
		Spot: reservation.Spot{
			ID:   0,
			Name: "sample-spot-name",
		},
	}

	// when
	output := formatter.FormatUnbookResponse(res)

	// assert
	snaps.MatchSnapshot(t, output)
}

func TestDiscordFormatter_FormatBookError(t *testing.T) {
	// given
	formatter := NewFormatter()
	response := book.BookResponse{
		Request: &book.BookRequest{
			Guild:          nil,
			Member:         nil,
			Spot:           "",
			StartAt:        time.Time{},
			EndAt:          time.Time{},
			Overbook:       false,
			HasPermissions: false,
		},
		ConflictingReservations: []*reservation.ClippedOrRemovedReservation{
			{
				Original: &reservation.Reservation{
					ID:      0,
					Author:  "sample-author",
					StartAt: time.Date(2021, 1, 1, 0, 0, 0, 0, time.UTC),
					EndAt:   time.Date(2021, 1, 1, 2, 0, 0, 0, time.UTC),
				},
				New: []*reservation.Reservation{
					{
						ID:      0,
						Author:  "sample-author",
						StartAt: time.Date(2021, 1, 1, 0, 0, 0, 0, time.UTC),
						EndAt:   time.Date(2021, 1, 1, 1, 30, 0, 0, time.UTC),
					},
				},
			},
		},
	}

	// when
	output := formatter.FormatBookError(response, errors.New("test error"))

	// assert
	snaps.MatchSnapshot(t, output)
}

func TestDiscordFormatter_FormatUnbookResponse1(t *testing.T) {
	// given
	formatter := NewFormatter()
	res := &reservation.ReservationWithSpot{
		Reservation: reservation.Reservation{
			ID:      0,
			SpotID:  0,
			StartAt: time.Date(2021, 1, 1, 0, 0, 0, 0, time.UTC),
			EndAt:   time.Date(2021, 1, 1, 2, 0, 0, 0, time.UTC),
		},
		Spot: reservation.Spot{
			ID:   0,
			Name: "test-spot",
		},
	}

	// when
	output := formatter.FormatUnbookResponse(res)

	// assert
	snaps.MatchSnapshot(t, output)
}

func TestDiscordFormatter_FormatBookResponse(t *testing.T) {
	// given
	formatter := NewFormatter()
	response := book.BookResponse{
		Request: &book.BookRequest{
			Member: &member.Member{
				ID:   "test-id",
				Nick: "test-nick",
			},
			Guild:   &guild.Guild{},
			Spot:    "test-spot",
			StartAt: time.Date(2021, 1, 1, 0, 0, 0, 0, time.UTC),
			EndAt:   time.Date(2021, 1, 1, 2, 0, 0, 0, time.UTC),
		},
		ConflictingReservations: []*reservation.ClippedOrRemovedReservation{
			{Original: &reservation.Reservation{
				ID:      0,
				Author:  "sample-author",
				StartAt: time.Date(2021, 1, 1, 0, 0, 0, 0, time.UTC),
				EndAt:   time.Date(2021, 1, 1, 2, 0, 0, 0, time.UTC),
			},
				New: []*reservation.Reservation{
					{
						ID:      0,
						Author:  "sample-author",
						StartAt: time.Date(2021, 1, 1, 0, 0, 0, 0, time.UTC),
						EndAt:   time.Date(2021, 1, 1, 1, 30, 0, 0, time.UTC),
					},
				},
			},
		},
	}

	// when
	output := formatter.FormatBookResponse(response)

	// assert
	snaps.MatchSnapshot(t, output)
}

func TestDiscordFormatter_FormatOverbookedMemberNotification(t *testing.T) {
	// given
	formatter := NewFormatter()
	member := &member.Member{
		ID:   "test-id",
		Nick: "test-nick",
	}
	request := book.BookRequest{
		Member: member,
	}
	res := &reservation.ClippedOrRemovedReservation{
		Original: &reservation.Reservation{
			ID:      0,
			StartAt: time.Date(2021, 1, 1, 0, 0, 0, 0, time.UTC),
			EndAt:   time.Date(2021, 1, 1, 2, 0, 0, 0, time.UTC),
		},
		New: []*reservation.Reservation{
			{
				ID:      1,
				StartAt: time.Date(2021, 1, 1, 0, 0, 0, 0, time.UTC),
				EndAt:   time.Date(2021, 1, 1, 1, 10, 0, 0, time.UTC),
			},
			{
				ID:      2,
				StartAt: time.Date(2021, 1, 1, 1, 30, 0, 0, time.UTC),
				EndAt:   time.Date(2021, 1, 1, 2, 0, 0, 0, time.UTC),
			},
		},
	}

	// when
	output := formatter.FormatOverbookedMemberNotification(member, request, res)

	// assert
	snaps.MatchSnapshot(t, output)
}
