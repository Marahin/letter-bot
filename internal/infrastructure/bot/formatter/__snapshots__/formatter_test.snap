
[TestDiscordFormatter_FormatGenericError - 1]
Sorry, but something went wrong. If you require support, join TibiaLoot.com Discord: https://discord.gg/F4YKgsnzmc 
Error message:
```
test error
```
---

[TestDiscordFormatter_FormatUnbookResponse - 1]
sample-spot-name (2021-01-01 00:00 - 2021-01-01 02:00) reservation has been cancelled.
---

[TestDiscordFormatter_FormatBookError - 1]
Sorry, but something went wrong. If you require support, join TibiaLoot.com Discord: https://discord.gg/F4YKgsnzmc 
Error message:
```
test error
```
Following reservations are conflicting:

* **sample-author** 2021-01-01 00:00 - 2021-01-01 02:00

---

[TestDiscordFormatter_FormatUnbookResponse1 - 1]
test-spot (2021-01-01 00:00 - 2021-01-01 02:00) reservation has been cancelled.
---

[TestDiscordFormatter_FormatBookResponse - 1]
<@!test-id> booked **test-spot** between 2021-01-01 00:00 and 2021-01-01 02:00.

Following reservations are conflicting **and have been shortened or removed**:

* <@!> had their reservation clipped to: **2021-01-01 00:00 - 2021-01-01 01:30** (originally: 2021-01-01 00:00 - 2021-01-01 02:00)

---

[TestDiscordFormatter_FormatOverbookedMemberNotification - 1]
Your reservation was overbooked by <@!test-id>
* <@!test-id>  has been clipped to: 2021-01-01 00:00 - 2021-01-01 01:10, 2021-01-01 01:30 - 2021-01-01 02:00
---
